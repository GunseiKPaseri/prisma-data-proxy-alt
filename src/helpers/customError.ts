import { Request, Response, NextFunction } from "express";

export const customError = (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  const oldSend = res.send;

  res.send = function (json) {
    try {
      const data = JSON.parse(json);
      const message = data.errors?.[0]?.message;
      res.send = oldSend;
      if (message) {
        res.statusCode = 400;
        return res.send(
          JSON.stringify({
            EngineNotStarted: {
              reason: {
                KnownEngineStartupError: {
                  msg: message,
                  error_code: "P5006",
                },
              },
            },
          })
        );
      }
      return res.send(json);
    } catch {
      res.send = oldSend;
      return res.send(json);
    }
  };

  next();
};
