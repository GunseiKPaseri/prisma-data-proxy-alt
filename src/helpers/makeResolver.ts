import { PrismaClient } from "@prisma/client/scripts/default-index";
import { DMMF } from "@prisma/client/runtime";
import { SelectionNode } from "graphql";
import { IFieldResolver } from "@graphql-tools/utils";

const toLowerFirstLetter = (str: string) => {
  return str.charAt(0).toLowerCase() + str.substring(1);
};

const functionsMappings = (
  dmmf: DMMF.Document
): Record<string, { operation: string; model: string }> => {
  return Object.fromEntries(
    dmmf.mappings.modelOperations.flatMap(
      ({ model, plural, ...operations }) => {
        return Object.entries(operations).map(([k, v]) => [
          v,
          { operation: k, model },
        ]);
      }
    )
  );
};

interface AggregateField {
  [x: string]: true | AggregateField;
}

const digAggregateField = (
  selections: readonly SelectionNode[],
  isRoot: boolean = true
): AggregateField => {
  return selections.reduce((res, selection) => {
    if (
      "name" in selection &&
      "selectionSet" in selection &&
      !(isRoot && !selection.name.value.startsWith("_"))
    ) {
      const dug = digAggregateField(
        selection.selectionSet?.selections ?? [],
        false
      );
      return {
        ...res,
        [selection.name.value]: Object.keys(dug).length ? dug : true,
      };
    }
    return res;
  }, {});
};

const rootOperationProxy = (db: PrismaClient, dmmf: DMMF.Document) => {
  const mapping = functionsMappings(dmmf);
  return new Proxy(
    {},
    {
      get: (_, method): IFieldResolver<Record<string, unknown>, unknown> => {
        return async (...[, args, , info]) => {
          const { operation, model } = mapping[method.toString()];

          const newArgs =
            operation === "aggregate" || operation === "groupBy"
              ? {
                  ...digAggregateField(
                    info.fieldNodes[0].selectionSet?.selections ?? []
                  ),
                  ...args,
                }
              : args;

          try {
            return db[toLowerFirstLetter(model)][operation](newArgs);
          } catch (e) {
            console.error(e);
            throw e;
          }
        };
      },
    }
  );
};

const relatedOperationProxy = (
  db: PrismaClient,
  { name, fields }: DMMF.Model
) => {
  return new Proxy(
    {},
    {
      get: (
        _,
        method
      ): IFieldResolver<Record<string, unknown>, unknown> | undefined => {
        const idFieldName = fields.find(({ isId }) => isId)?.name;
        const field = fields.find(({ name }) => name === method.toString());
        if (!idFieldName || !field?.relationName) return undefined;

        return async (parent, args) => {
          try {
            return await db[toLowerFirstLetter(name)]
              .findUnique({ where: { [idFieldName]: parent[idFieldName] } })
              [method](args);
          } catch (e) {
            console.error(e);
            throw e;
          }
        };
      },
    }
  );
};

export const makeResolver = (
  db: PrismaClient,
  { dmmf }: { dmmf: DMMF.Document }
) => {
  return {
    ...Object.fromEntries(
      dmmf.datamodel.models.map((model) => [
        model.name,
        relatedOperationProxy(db, model),
      ])
    ),
    Query: rootOperationProxy(db, dmmf),
    Mutation: rootOperationProxy(db, dmmf),
  };
};
