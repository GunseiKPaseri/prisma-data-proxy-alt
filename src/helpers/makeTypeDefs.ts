import { Prisma } from "@prisma/client";
import { gql } from "apollo-server-express";

const Definitions = [
  ...Prisma.dmmf.schema.outputObjectTypes.prisma,
  ...Prisma.dmmf.schema.outputObjectTypes.model,
].map((m) => {
  return `
      type ${m.name} {
        ${m.fields
          .map((f) => {
            return `${f.name}${
              f.args.length > 0
                ? `(${f.args.map(({ name }) => `${name}: Any`).join(" ")})`
                : ""
            }: ${
              f.outputType.isList
                ? `[${f.outputType.type}]`
                : `${f.outputType.type}`
            }${f.isNullable ? "!" : ""}`;
          })
          .join(" ")}
      }
      `;
});

export const typeDefs = gql`
  scalar Any
  scalar DateTime
  scalar Json

  ${Definitions}
`;
