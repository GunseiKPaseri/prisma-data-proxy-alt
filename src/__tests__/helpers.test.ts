import { describe, expect, test, vi } from "vitest";
import { getSampleDMMF } from "./__fixtures__/getSampleSchema";
import { makeTypeDefs } from "../helpers/makeTypeDefs";
import {
  rootOperationProxy,
  relatedOperationProxy,
} from "../helpers/makeResolver";
import { parse } from "graphql";

describe("makeTypeDefs", () => {
  test("Check the generated graphql schema.", async () => {
    const dmmf = await getSampleDMMF();
    const typeDef = makeTypeDefs({ dmmf });

    expect(typeDef.loc?.source.body).toMatchSnapshot();
  });
});

describe("makeResolver > rootOperationProxy", () => {
  test("called fineManyUser", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const proxy = rootOperationProxy(
      {
        user: {
          findMany: mock,
        },
      },
      dmmf
    );

    // @ts-ignore
    proxy.findManyUser({}, { foo: "foo", bar: "bar" });
    expect(mock).toBeCalledWith({ foo: "foo", bar: "bar" });
  });

  test("called aggregateUser", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const proxy = rootOperationProxy(
      {
        user: {
          aggregate: mock,
        },
      },
      dmmf
    );

    const a = parse(`query {
      aggregateUser(where: {
        id: 1
      }) {
        _count {
          _all
          id
        }
      }
    }`);

    // @ts-ignore
    proxy.aggregateUser(
      {},
      {
        where: {
          id: 1,
        },
      },
      null,
      {
        // @ts-ignore
        fieldNodes: a.definitions[0].selectionSet.selections,
      }
    );
    expect(mock).toBeCalledWith({
      where: {
        id: 1,
      },
      _count: { _all: true, id: true },
    });
  });

  test("called groupByUser", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const proxy = rootOperationProxy(
      {
        user: {
          groupBy: mock,
        },
      },
      dmmf
    );

    const a = parse(`query {
      groupByUser(by: [name]) {
        _max {
          id
        }
        name
      }
    }`);

    // @ts-ignore
    proxy.groupByUser({}, { by: ["name"] }, null, {
      // @ts-ignore
      fieldNodes: a.definitions[0].selectionSet.selections,
    });
    expect(mock).toBeCalledWith({
      by: ["name"],
      _max: { id: true },
    });
  });
});

describe("makeResolver > relatedOperationProxy", () => {
  test("called user.Team", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const relatedMock = vi.fn();
    const proxy = relatedOperationProxy(
      {
        user: {
          findUnique: mock.mockReturnValue({ Team: relatedMock }),
        },
      },
      dmmf.datamodel.models.find(({ name }) => name === "User")!
    );

    // @ts-ignore
    proxy.Team({ id: 10, name: "foo" }, { select: { id: true, name: true } });
    expect(mock).toBeCalledWith({ where: { id: 10 } });
    expect(relatedMock).toBeCalledWith({ select: { id: true, name: true } });
  });
});
