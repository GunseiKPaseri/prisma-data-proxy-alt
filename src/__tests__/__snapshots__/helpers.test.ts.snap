// Vitest Snapshot v1

exports[`makeTypeDefs > Check the generated graphql schema. 1`] = `
"
    scalar Any
    scalar DateTime
    scalar Json
    
      enum UserScalarFieldEnum {
        id email name teamId createdAt updatedAt
      }
    ,
      enum TeamScalarFieldEnum {
        id name language createdAt updatedAt
      }
    ,
      enum SortOrder {
        asc desc
      }
    ,
      enum QueryMode {
        default insensitive
      }
    ,
      enum Language {
        JAPANESE ENGLISH CHINESE FRENCH GERMAN SPANISH
      }
    
    
      type Query {
        findFirstUser(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): User findManyUser(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): [User]! aggregateUser(where: Any orderBy: Any cursor: Any take: Any skip: Any): AggregateUser! groupByUser(where: Any orderBy: Any by: Any having: Any take: Any skip: Any): [UserGroupByOutputType]! findUniqueUser(where: Any): User findFirstTeam(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): Team findManyTeam(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): [Team]! aggregateTeam(where: Any orderBy: Any cursor: Any take: Any skip: Any): AggregateTeam! groupByTeam(where: Any orderBy: Any by: Any having: Any take: Any skip: Any): [TeamGroupByOutputType]! findUniqueTeam(where: Any): Team
      }
      ,
      type Mutation {
        createOneUser(data: Any): User! upsertOneUser(where: Any create: Any update: Any): User! createManyUser(data: Any skipDuplicates: Any): AffectedRowsOutput! deleteOneUser(where: Any): User updateOneUser(data: Any where: Any): User updateManyUser(data: Any where: Any): AffectedRowsOutput! deleteManyUser(where: Any): AffectedRowsOutput! createOneTeam(data: Any): Team! upsertOneTeam(where: Any create: Any update: Any): Team! createManyTeam(data: Any skipDuplicates: Any): AffectedRowsOutput! deleteOneTeam(where: Any): Team updateOneTeam(data: Any where: Any): Team updateManyTeam(data: Any where: Any): AffectedRowsOutput! deleteManyTeam(where: Any): AffectedRowsOutput! executeRaw(query: Any parameters: Any): Json! queryRaw(query: Any parameters: Any): Json!
      }
      ,
      type AggregateUser {
        _count: UserCountAggregateOutputType _avg: UserAvgAggregateOutputType _sum: UserSumAggregateOutputType _min: UserMinAggregateOutputType _max: UserMaxAggregateOutputType
      }
      ,
      type UserGroupByOutputType {
        id: Int! email: String! name: String teamId: String createdAt: DateTime! updatedAt: DateTime! _count: UserCountAggregateOutputType _avg: UserAvgAggregateOutputType _sum: UserSumAggregateOutputType _min: UserMinAggregateOutputType _max: UserMaxAggregateOutputType
      }
      ,
      type AggregateTeam {
        _count: TeamCountAggregateOutputType _min: TeamMinAggregateOutputType _max: TeamMaxAggregateOutputType
      }
      ,
      type TeamGroupByOutputType {
        id: String! name: String! language: [Language] createdAt: DateTime! updatedAt: DateTime! _count: TeamCountAggregateOutputType _min: TeamMinAggregateOutputType _max: TeamMaxAggregateOutputType
      }
      ,
      type AffectedRowsOutput {
        count: Int!
      }
      ,
      type UserCountAggregateOutputType {
        id: Int! email: Int! name: Int! teamId: Int! createdAt: Int! updatedAt: Int! _all: Int!
      }
      ,
      type UserAvgAggregateOutputType {
        id: Float
      }
      ,
      type UserSumAggregateOutputType {
        id: Int
      }
      ,
      type UserMinAggregateOutputType {
        id: Int email: String name: String teamId: String createdAt: DateTime updatedAt: DateTime
      }
      ,
      type UserMaxAggregateOutputType {
        id: Int email: String name: String teamId: String createdAt: DateTime updatedAt: DateTime
      }
      ,
      type TeamCountOutputType {
        users: Int!
      }
      ,
      type TeamCountAggregateOutputType {
        id: Int! name: Int! language: Int! createdAt: Int! updatedAt: Int! _all: Int!
      }
      ,
      type TeamMinAggregateOutputType {
        id: String name: String createdAt: DateTime updatedAt: DateTime
      }
      ,
      type TeamMaxAggregateOutputType {
        id: String name: String createdAt: DateTime updatedAt: DateTime
      }
      ,
      type User {
        id: Int! email: String! name: String teamId: String createdAt: DateTime! updatedAt: DateTime! Team: Team
      }
      ,
      type Team {
        id: String! name: String! language: [Language] users(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): [User] createdAt: DateTime! updatedAt: DateTime! _count: TeamCountOutputType!
      }
      
  "
`;
